### DISCUSSION TITLE
Configuring two Orchard Core instances to exchange data using APIs with OAuth 2.0 `Client Credentials Flow`

### DISCUSSION THREAD

**Discussion by MarGraz** [OrchardCMS/OrchardCore#16138] [Q&A]
Hi everyone,

I'm curious to know if any of you have already implemented two Orchard Core web apps that "talk" to each other using custom APIs.

I'm implementing two microservices that will "talk" to each other, and I want to use the `OpenId` module. Of course, I need the `Client` and `Server` modules installed, but I'm curious to know if there's any public example on how to implement this, to learn best practices in Orchard Core, and if you have implemented your own `OAuthService` to manage the `GetAccessToken` method in the OAuth 2.0 [Client Credentials Flow](https://documentation.openiddict.com/guides/choosing-the-right-flow.html#client-credentials-grant-recommended-for-machine-to-machine-communication) (this is recommended for machine-to-machine communication).

My questions are:

1. Are there any public examples on how to implement this to learn best practices in Orchard Core?
2. Have you implemented your own `OAuthService` to manage the `GetAccessToken` method?

Thank you 😊

**Comment by douwinga** [comment]
I am also curious to know other peoples solution to this. What we did is implement Azure Service Bus. We just created custom workflow tasks to send service bus messages as well as to subscribe to topics. Therefore, one instance can send a message with content on it and then any other instance can subscribe to the topic if they want to receive those content updates. This works for us, but the implementation does have some issues as saving a subscription workflow does a tenant reload so that its subscription gets updated and this causes issues sometimes when other things like lucene are updating the index or something.

**Comment by Piedone** [comment]
Not an answer, but you may take some inspiration from https://github.com/Lombiq/Orchard-Core-API-Client.

**Reply by MarGraz** [reply]
Thank you @Piedone. 

I noticed that the "OC API Client" uses a certificate to validate calls. I think the [ConfigurableCertificateValidatingHttpClientHandler](https://github.com/Lombiq/Orchard-Core-API-Client/blob/dev/Lombiq.OrchardCoreApiClient/ConfigurableCertificateValidatingHttpClientHandler.cs) class could serve as good inspiration. I will remove the certificate part to use only the "Client credentials flow", and it should be okay.

I will keep you posted. If you hear a boom, it’s probably me 🤣

**Reply by Piedone** [reply]
Sorry about that :D.
